### DISCUSSION TITLE
Deno 2.0 Decorators

### DISCUSSION THREAD

**Discussion by cfjello** [denoland/deno#26737] [Q&A]
I have some older decorators with parameters that used to work in Deno/Typescript, here illustrated with a simple example:
```
function PluginDecorator(name: string) {
    return (_ctor: () => void) => {
        console.log("Plugin found: " + name);
    }
}

@PluginDecorator("My first Plugin")
class myFirstPlugin {
   
}
```
It now gives me an error:
```
Unable to resolve signature of class decorator when called as an expression.
  The runtime will invoke the decorator with 2 arguments, but the decorator expects 1.deno-ts(1238)
```
In Deno 2.0 the following options are depreciated:
```
{
    "compilerOptions": {
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true 
    }
}
```
and now gives me ```Value is deprecated```, and I have not been able to find a work around for this. Can somebody please advise, it would much appreciated, best of all if you could provide me with a small working example?

**Comment by bartlomieju** [comment]
With the `Value is deprecated` warning does the decorator work as expected?

**Reply by cfjello** [reply]
It does not work, unfortunately. I investigated a little bit further using Typescript tsc instead of Deno, with and without the flags, but it also errors, so it seems to be an issue originating in Typescript. I will opt for writing some mixin-functions instead.

**Comment by nathanwhit** [comment]
I don't fully understand why, or when this changed, but the issue is with the typing of the argument to your decorator. You have it typed as `() => void`, but typescript will be passing a [class type](https://www.typescriptlang.org/docs/handbook/2/generics.html#using-class-types-in-generics) to it. Instead you should type it as `{ new (): void }` or `new () => void`.

So

```ts
function PluginDecorator(name: string) {
    return (_ctor: new () => void) => { // note the insertion of `new`
        console.log("Plugin found: " + name);
    }
}

@PluginDecorator("My first Plugin")
class myFirstPlugin {
   
}
```

This works, but as per the example in the [typescript reference](https://www.typescriptlang.org/docs/handbook/decorators.html#class-decorators) on class decorators (it's the example at the end of the `Class Decorators` section) I think the most correct typing would be more like

```ts
function PluginDecorator(name: string) {
    return <T extends { new (...args: any[]): {} }>(_ctor: T) => {
        console.log("Plugin found: " + name);
    }
}
```

**Reply by cfjello** [reply]
Great! They both work without any  flags in my setup with the newest Deno v.2.0..5, even though the vscode Deno linter still complans: "Unable to resolve signature of class decorator ....". Thanks!
